File: 23. UniqueSubsetsOfString.py
# Topic: Find all unique subsets of a string using bits

def unique_subsets(s):
    n = len(s); seen = set(); res = []
    for mask in range(1<<n):
        cur = ''.join(s[i] for i in range(n) if mask & (1<<i))
        if cur not in seen:
            seen.add(cur); res.append(cur)
    return res

if __name__ == '__main__':
    print('Unique subsets count:', len(unique_subsets('aba')))

# Time Complexity: O(2^n * n)
# Space Complexity: O(2^n * n)

File: 23. UniqueSubsetsOfString.cpp
// Topic: Find all unique subsets of a string using bits
#include <bits/stdc++.h>
using namespace std;

vector<string> uniqueSubsets(const string& s)
{
    int n = s.size();
    set<string> seen;
    vector<string> res;
    for (int mask = 0; mask < (1 << n); ++mask)
    {
        string cur;
        for (int i = 0; i < n; ++i)
            if (mask & (1 << i))
                cur.push_back(s[i]);
        if (seen.insert(cur).second)
            res.push_back(cur);
    }
    return res;
}

int main()
{
    auto u = uniqueSubsets("aba");
    cout << "Unique subsets: " << u.size() << '\n';
    return 0;
}

// Time Complexity: O(2^n * n)
// Space Complexity: O(2^n * n)
